feat: add multiple tags in bulk for a monitor

grvwy opened the pull request:
## 📋 Overview

- **What problem does this pull request address?**

  - A personal problem when adding tags, if you create a lot of monitors and groups it gets tiresome only being able to add a tag one at the time. Especially if you have monitor or group specific tags.

- **What features or functionality does this pull request introduce or enhance?**

This improves the TagManager.vue modal with being able to:
 
  - **Tag Staging:** Users can define a tag (either by selecting an existing system tag and providing a value, or by creating a new system tag with a name, color, and value) and "stage" it for addition without closing the modal.
    - **Staging Area:** A new display area within the modal shows a list of all tags currently staged for addition.
    - **Batch Addition:** Users can stage multiple tags and then commit them all at once to the monitor.
    - **Individual Unstaging:** Tags can be removed individually from the staged list before final confirmation.
    - **"Clear Form" Control:** A link to quickly clear the current tag input fields without affecting staged tags or closing the modal, useful for correcting input errors or discarding a draft.
    - **Streamlined Single Tag Addition:** If a user defines a single valid tag and clicks the main "Add" button in the footer (without explicitly staging via "Add Another Tag"), this single tag is directly added.
    - **Validation Feedback:** Improved validation messages for various scenarios (tag limit, duplicates, etc.).
    - **New System Tag Limit:** A configurable constant (`MAX_NEW_SYSTEM_TAGS_PER_BATCH`) limits the number of *newly created system tags* that can be added in one batch.

## 🔄 Changes

### 🛠️ Type of change

<!-- Please select all options that apply -->

- [ ] 🐛 Bugfix (a non-breaking change that resolves an issue)
- [X] ✨ New feature (a non-breaking change that adds new functionality)
- [ ] ⚠️ Breaking change (a fix or feature that alters existing functionality in a way that could cause issues)
- [X] 🎨 User Interface (UI) updates
- [ ] 📄 New Documentation (addition of new documentation)
- [ ] 📄 Documentation Update (modification of existing documentation)
- [ ] 📄 Documentation Update Required (the change requires updates to related documentation)
- [ ] 🔧 Other (please specify):
  - Provide additional details here.

## 🔗 Related Issues

- Resolves https://github.com/louislam/uptime-kuma/issues/5849

## 📄 Checklist *

<!-- Please select all options that apply -->

- [X] 🔍 My code adheres to the style guidelines of this project.
- [X] ✅ I ran ESLint and other code linters for modified files.
- [X] 🛠️ I have reviewed and tested my code.
- [X] 📝 I have commented my code, especially in hard-to-understand areas (e.g., using JSDoc for methods).
- [X] ⚠️ My changes generate no new warnings.
- [ ] 🤖 My code needed automated testing. I have added them (this is an optional task).
- [ ] 📄 Documentation updates are included (if applicable).
- [X] 🔒 I have considered potential security impacts and mitigated risks.
- [X] 🧰 Dependency updates are listed and explained.
- [X] 📚 I have read and understood the [Pull Request guidelines](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#recommended-pull-request-guideline).

## 📷 Screenshots or Visual Changes

![image](https://github.com/user-attachments/assets/dba4fd20-5c84-4d76-8a07-fc722e8e4b60)
![image](https://github.com/user-attachments/assets/c3e82a8e-bd76-4957-8b3c-5a2a9f1ecedf)
![image](https://github.com/user-attachments/assets/74d0e390-6e81-45b1-abc5-4167837ba2be)


- **UI Modifications**:
    - The "Add Tag" modal in `TagsManager.vue` has been significantly redesigned for batch tag additions.
    - **Before**: The modal allowed adding only one tag at a time. The user would select or define a tag, click "Add", and the modal would close.
    - **After**:
        - A new "Staging Area" is present at the top of the modal, displaying tags as they are prepared for batch addition.
        - The primary button in the modal body has changed from a simple "Add" (which closed the modal) to an "Add Another Tag" button. This button stages the currently defined tag and clears the input fields, allowing the user to define the next tag.
        - A "Clear Form" link has been added next to the "Add Another Tag" button, allowing users to reset the input fields without closing the modal or losing already staged tags.
        - The modal footer now contains a "Cancel" button (to discard all staged tags and close) and an "Add" button (to commit all staged tags to the monitor).
        - If only one tag is defined in the input fields (and nothing is staged), the footer "Add" button can be used to add that single tag directly, streamlining the process for single additions.

## ℹ️ Additional Context

Provide any relevant details to assist reviewers in understanding the changes.

<details><summary>Click here for more details:</summary>
</p>

**Key Considerations**:

- **Design decisions** – Key choices or trade-offs made during development.
- **Alternative solutions** – Approaches considered but not implemented, along with reasons.
- **Relevant links** – Specifications, discussions, or resources that provide context.
- **Dependencies** – Related pull requests or issues that must be resolved before merging.
- **Additional context** – Any other details that may help reviewers understand the changes.

Provide details here

## 💬 Requested Feedback

<!-- If a part of our docs is unclear, you are unsure how to do something/.. this is where we would appreciate your feedback -->

- `Mention documents needing feedback here`

CommanderStorm commented on .gitignore:
```diff
@@ -18,6 +18,7 @@ dist-ssr
 /extra/healthcheck.exe
 /extra/healthcheck
 /extra/healthcheck-armv7
+/studies
```
This is not one of our directories..

We can add this if it is something common somewhere/..


```suggestion
```

grvwy replied:
That was accidental from my branch, sorry!

CommanderStorm commented on src/components/TagsManager.vue:
```diff
@@ -104,34 +113,45 @@
 </template>
 
 <script>
+const MAX_NEW_SYSTEM_TAGS_PER_BATCH = 10; // Example limit
```
Lets not limit this.. I think removing this part entirely is fine.

grvwy replied:
Are you sure? I was thinking limiting was just for good measure and to have pre-implemented easy restraint if needed for future. I haven't tested what happens to the system when adding an abnormal amount of tags

CommanderStorm commented on src/components/TagsManager.vue:
```diff
@@ -176,13 +196,17 @@ export default {
                 value: "",
-                invalid: true,
-                nameInvalid: false,
+                // invalid: true, // Initial validation will be handled by computed prop
```
```suggestion
```

grvwy replied:
Agreed, redundant

grvwy commented:
> Here is a bit of intial feedback about the design.
> 
> ![445350889-dba4fd20-5c84-4d76-8a07-fc722e8e4b60](https://private-user-images.githubusercontent.com/26258709/445705514-68a5858b-1ea2-4c9b-848d-cf27fa24fb67.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc3Njk1NzQsIm5iZiI6MTc0Nzc2OTI3NCwicGF0aCI6Ii8yNjI1ODcwOS80NDU3MDU1MTQtNjhhNTg1OGItMWVhMi00YzliLTg0OGQtY2YyN2ZhMjRmYjY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTIwVDE5Mjc1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU5ODc1MTIxMGUzMmIwOTg0ODI0OGYzMDk0ZDNiODc5MzE4ZWRhNDYyZGQ2ODgwOThhZTFlMTg2NmM3ZTE2YWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.J_1TS4LF3VMAHfIVnyUwXcQlrUXug-FvHb7N28kazHs)
> 
> Could you please make sure to add a testcase that this works as intended here? https://github.com/louislam/uptime-kuma/blob/acf724412a75105ecf06779752195abf5ba8ce4e/test/e2e/specs/status-page.spec.js


what about this?
![image](https://github.com/user-attachments/assets/39b445a3-82e9-4591-91d7-612b1081ea20)

CommanderStorm commented:
I don't think the clear form button is nesssesary.
And if we remove it, we can move the add more tags button next to the save/add tags button.

Sorry, above annotations might have been not very clear

grvwy commented:
> I don't think the clear form button is nesssesary. And if we remove it, we can move the add more tags button next to the save/add tags button.
> 
> Sorry, above annotations might have been not very clear

![image](https://github.com/user-attachments/assets/95d9d7c9-8f6c-4046-a8f8-ae44a494beb5)

are we moving the right way?

grvwy commented:
- UI updated:
![image](https://github.com/user-attachments/assets/3ec12553-a7b7-482f-b429-23e3037127d4)

- tests added to status-page.spec.js to verify multi-tag functions

- I opted to keep 
> Lets not limit this.. I think removing this part entirely is fine.

const MAX_NEW_SYSTEM_TAGS_PER_BATCH = 10; // Example limit 

It just feels good to have in place, but if you're adamant I'll buy it. 😄

CommanderStorm commented:
Regarding the limit:
The main reason why I asked you to remove it is that I suspect that it in practice is dead code.
Who has more than 10 tags? At this point, that is a bit excessive. 1-4, yes. 10+? unlikely.

Our existing UI has not been tested for this and likely breaks down somewhere in ugly ways.

=> lets remove this code until we identify a need for such code.

CommanderStorm commented on src/components/TagsManager.vue:
```diff
@@ -104,34 +113,46 @@
                     </div>
+                    <!-- Modal Footer: Cancel batch or Confirm and Add all staged tags -->
+                    <div class="modal-footer">
+                        <button type="button" class="btn btn-outline-primary me-2" :disabled="processing || validateDraftTag.invalid" @click.stop="stageCurrentTag">
+                            {{ $t("Add Another Tag") }}
+                        </button>
+                        <button type="button" class="btn btn-secondary" @click.stop="clearStagingAndCloseModal">{{ $t("Cancel") }}</button>
```
not sure, but I think if we order it cancel, add another´, add all this seems more "natural" to me.
Not sure why, likely just being used to this.

```suggestion
                        <button type="button" class="btn btn-secondary" @click.stop="clearStagingAndCloseModal">{{ $t("Cancel") }}</button>
                        <button type="button" class="btn btn-outline-primary me-2" :disabled="processing || validateDraftTag.invalid" @click.stop="stageCurrentTag">
                            {{ $t("Add Another Tag") }}
                        </button>
```

CommanderStorm commented on src/lang/en.json:
```diff
@@ -1098,5 +1102,9 @@
-    "Disable URL in Notification": "Disable URL in Notification"
+    "Disable URL in Notification": "Disable URL in Notification",
+    "Add Another Tag": "Add Another Tag",
+    "Staged Tags for Batch Add": "Staged Tags for Batch Add",
```
also please make a pass that all translations are listed in this file.
If a key (like `add tags`) is not here, it is not translatable into other languages by our kickass translators 😉

grvwy commented:
I hope this satisfactory

tests added to verify adding multiple tags

moved the "add another tag" to the footer and re-organized

![image](https://github.com/user-attachments/assets/58b79c37-9823-47d6-951c-2db06a10b3dc)

CommanderStorm commented on server/model/status_page.js:
```diff
@@ -120,8 +120,8 @@ class StatusPage extends BeanModel {
 
         const head = $("head");
 
-        if (statusPage.googleAnalyticsTagId) {
```
This is new: Is this intentional?

grvwy replied:
Without it the test e2e test failed previously, lets try it out

louislam replied:
I think it should not be touch, as it should be working fine. And this pr should not related to status page?

CommanderStorm replied:
there was a testcase added for this, said testcase failed.
This change should be fine (even though: next time lets do it in a different PR ^^)

louislam commented:
1. Using `Done` is better?

![image](https://github.com/user-attachments/assets/922e6f52-1385-4acd-a4f0-258d163fcfe8)

2. Clicking outside the modal closes the modal. User could lose all tags in one accidentally.


3. Not sure if it is related to this pr, but remove a tag, also remove all same tags with different values.


https://github.com/user-attachments/assets/31a615b8-e18b-44a2-9ddc-40e9f026985d

grvwy commented:
I changed the button to "done" from "add tags", that was a good UX suggestion

![image](https://github.com/user-attachments/assets/13287588-cc0d-4ae0-93b1-d618b3c6c7c5)


I also fixed the bug with the removal, you can now remove tags with the same name but with different values individually 

Fixed the accidentally closing out the modal, user now has to select cancel or done

Also noticed a bug where if you add a tag, press "add another tag" and try to re-use the one you created it wasn't an option. It is now properly staged.

louislam commented on src/lang/en.json:
```diff
@@ -517,7 +522,7 @@
     "Clone": "Clone",
     "cloneOf": "Clone of {0}",
-    "smtp": "Email (SMTP)",
+    "smtp": "Email ('')",
```
I guess it should be a mistake.

CommanderStorm commented on src/lang/en.json:
```diff
@@ -188,9 +188,14 @@
     "Tag with this value already exist.": "Tag with this value already exists.",
+    "tagAlreadyOnMonitor": "This tag (name and value) is already on the monitor or pending addition.",
+    "tagAlreadyStaged": "This tag (name and value) is already staged for this batch.",
+    "tagLimitReached": "Cannot stage more than {limit} new system tags in a single batch.",
```
```suggestion
```

CommanderStorm commented on src/components/TagsManager.vue:
```diff
@@ -104,27 +113,34 @@
-                                {{ $t("Add") }}
-                            </button>
+
+                        <!-- General feedback area for tag input validation messages -->
+                        <div v-if="validateDraftTag.invalid && validateDraftTag.messageKey" class="form-text text-danger mb-2">
+                            {{ $t(validateDraftTag.messageKey, validateDraftTag.messageParams) }}
                         </div>
+                        <!-- End Validation Feedback -->
+
+                        <!-- Add Another Tag Button -->
+                        <!-- This button will be moved to the footer -->
+
+                        <!-- Action Buttons: Clear current form and Stage current tag -->
+                        <!-- Removed original action buttons here -->
                     </div>
+                    <!-- Modal Footer: Cancel batch or Confirm and Add all staged tags -->
+                    <div class="modal-footer">
+                        <button type="button" class="btn btn-secondary" @click.stop="clearStagingAndCloseModal">{{ $t("Cancel") }}</button>
+                        <button type="button" class="btn btn-outline-primary me-2" :disabled="processing || validateDraftTag.invalid" @click.stop="stageCurrentTag">
+                            {{ $t("Add Another Tag") }}
+                        </button>
+                        <button type="button" class="btn btn-primary" :disabled="processing || (stagedForBatchAdd.length === 0 && validateDraftTag.invalid)" data-testid="add-tags-final-button" @click.stop="confirmAndCommitStagedTags">{{ $t("Done") }}</button>
+                    </div>
+                    <!-- End Modal Footer -->
```
```suggestion
                        <div v-if="validateDraftTag.invalid && validateDraftTag.messageKey" class="form-text text-danger mb-2">
                            {{ $t(validateDraftTag.messageKey, validateDraftTag.messageParams) }}
                        </div>
                    </div>
                    <div class="modal-footer">
                        <button type="button" class="btn btn-secondary" @click.stop="clearStagingAndCloseModal">{{ $t("Cancel") }}</button>
                        <button type="button" class="btn btn-outline-primary me-2" :disabled="processing || validateDraftTag.invalid" @click.stop="stageCurrentTag">
                            {{ $t("Add Another Tag") }}
                        </button>
                        <button type="button" class="btn btn-primary" :disabled="processing || (stagedForBatchAdd.length === 0 && validateDraftTag.invalid)" data-testid="add-tags-final-button" @click.stop="confirmAndCommitStagedTags">{{ $t("Done") }}</button>
                    </div>
```

CommanderStorm commented on src/components/TagsManager.vue:
```diff
@@ -20,10 +20,22 @@
             <div class="modal-dialog modal-dialog-centered">
                 <div class="modal-content">
                     <div class="modal-body">
+                        <!-- Display area for tags staged for batch addition -->
+                        <h4 v-if="stagedForBatchAdd.length > 0">{{ $t("Add Tags") }}</h4>
+                        <div v-if="stagedForBatchAdd.length > 0" class="mb-3 staging-area" style="max-height: 150px; overflow-y: auto;">
+                            <Tag
+                                v-for="stagedTag in stagedForBatchAdd"
+                                :key="stagedTag.keyForList"
+                                :item="mapStagedTagToDisplayItem(stagedTag)"
+                                :remove="() => unstageTag(stagedTag)"
+                            />
+                        </div>
+                        <!-- End Staging Display -->
```
```suggestion
                        <h4 v-if="stagedForBatchAdd.length > 0">{{ $t("Add Tags") }}</h4>
                        <div v-if="stagedForBatchAdd.length > 0" class="mb-3 staging-area" style="max-height: 150px; overflow-y: auto;">
                            <Tag
                                v-for="stagedTag in stagedForBatchAdd"
                                :key="stagedTag.keyForList"
                                :item="mapStagedTagToDisplayItem(stagedTag)"
                                :remove="() => unstageTag(stagedTag)"
                            />
                        </div>
```

CommanderStorm reviewed (approved):
Thanks for the enhancement ❤️
